### PR TITLE
feat: handshake and dc fork equivalence

### PR DESCRIPTION
**Pull request by luisacicolini**
This PR adds the first example containing the sketch of a proof for handshake and dc equivalence. Although very much experimental (we are not yet sure about the encoding of the dialects' semantics), we'd like to start merging the examples to avoid remaining too much behind main. cc: @ymherklotz @alexkeizer

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
We should merge early and often, as long as the code is tested and fails reliably, there is little benefit of maintaining branches.

**Comment by luisacicolini** [comment]
I polished the code a bit and the most important proof is complete, so I'd merge now. There's not much testing but I can add some to ci if necessary.
Let's maybe wait for an ack by @alexkeizer and @ymherklotz and then merge?

File: SSA/Projects/CIRCT/Handshake-to-DC/ForkData.lean [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        have h : ((fun s' => match
+            Id.run (match s' 0 with
+              | some _ => (s' 0, some (), s'.tail)
+              | none => (none, none, s'.tail)) with
+          | (_, a, b) => (a, b)) : Stream Int → Option Unit × Stream Int) =
+          (fun s' => (Option.map (fun _ => ()) (s' 0), s'.tail)) := by
+          ext s' <;> cases h : s' 0 <;> simp_all [Id.run]
+        unfold CIRCTStream.DC.unpack.match_1 CIRCTStream.Stream.Bisim.corec₂_corec1.match_1 CIRCTStream.Stream.Bisim.equiv_fork_fst.match_2 CIRCTStream.Stream.Bisim.equiv_fork_fst.match_1 at *
+        dsimp only [Id.run, TyDenote.toType] at *
+        set_option pp.explicit true in trace_state
```
Comment (ymherklotz) [inline SSA/Projects/CIRCT/Handshake-to-DC/ForkData.lean]: I would maybe remove this set_option at least.

**Review by ymherklotz — APPROVED** [review]
Looks good to me, would just clean-up the one set_option, but we can work on better cleaning up the proof in a later patch.

File: SSA/Projects/CIRCT/Handshake-to-DC/ForkData.lean (hunk 1 above)
Comment (luisacicolini) [inline SSA/Projects/CIRCT/Handshake-to-DC/ForkData.lean]:
```suggestion
```

File: SSA/Projects/CIRCT/DC/DC.lean [hunk 2]
```diff
@@ -125,8 +131,9 @@ instance instDCTyDenote : TyDenote Ty where
 toType := fun
 | Ty.tokenstream => CIRCTStream.DC.TokenStream
 | Ty.tokenstream2 => CIRCTStream.DC.TokenStream × CIRCTStream.DC.TokenStream
-| Ty.valuestream ty2 => CIRCTStream.DC.ValueStream (toType ty2)
-| Ty.valuetokenstream ty2 => CIRCTStream.DC.ValueStream (toType ty2) × CIRCTStream.DC.TokenStream
+| Ty.valuestream Ty2 => CIRCTStream.DC.ValueStream (toType Ty2)
+| Ty.valuestream2 Ty2 => CIRCTStream.DC.ValueStream (toType Ty2) × CIRCTStream.DC.ValueStream (toType Ty2)
+| Ty.valuetokenstream Ty2 => CIRCTStream.DC.ValueStream (toType Ty2) × CIRCTStream.DC.TokenStream
 
 set_option linter.dupNamespace false in
```
Comment (luisacicolini) [inline SSA/Projects/CIRCT/DC/DC.lean]:
```suggestion
```

**Pull request merged**